code/game/objects/items/devices/lightreplacer.dm

```diff
@@ -77,111 +115,204 @@
+
+	else
+		dat += "<h3>No supply container inserted. This should be impossible. Please ahelp this.</h3>"
+
+	if(waste)
+		dat += {"<br><br><br><h3>Waste Container:</h3>
+	
+		<b>Filled: </b>[waste.contents.len]/[waste.storage_slots]<br>
+		<b><a href='?src=\ref[src];recycle=1'>Recycle Contents</a></b>
+		"}
+	else
+		dat += "<br><br><br><h3>No waste container inserted. This should be impossible. Please ahelp this.</h3>"
+
+	var/datum/browser/popup = new(user, "lightreplacer", "", nref = src)
+	popup.set_content(dat)
+	popup.open()
 
 /obj/item/device/lightreplacer/update_icon()
 	icon_state = "lightreplacer[emagged]"
 
 
-/obj/item/device/lightreplacer/proc/Use(var/mob/user)
+/obj/item/device/lightreplacer/proc/ReplaceLight(var/obj/machinery/light/target, var/mob/living/user)
+	var/obj/item/weapon/light/best_light = get_best_light(target)
+	if(best_light == 0)
+		user << "<span class='warning'>\The [src] has no supply container!</span>"
+		return
+	else if(!best_light)
+		user << "<span class='warning'>\The [src] has no compatible light!</span>"
+		return
+	if(!is_light_better(best_light, target))
+		user << "<span class='notice'>\The [src] has no light better than the one already in \the [target].</span>"
+		return
 
 
+	user << "<span class='notice'>You replace the [target.fitting] with \the [src].</span>"
 	playsound(get_turf(src), 'sound/machines/click.ogg', 50, 1)
-	AddUses(-1)
-	return 1
 
-// Negative numbers will subtract
-/obj/item/device/lightreplacer/proc/AddUses(var/amount = 1)
-	uses = min(max(uses + amount, 0), max_uses)
-
-/obj/item/device/lightreplacer/proc/Charge(var/mob/user)
-	charge += 1
-	if(charge > 7)
-		AddUses(1)
-		charge = 1
-
-/obj/item/device/lightreplacer/proc/ReplaceLight(var/obj/machinery/light/target, var/mob/living/U)
-
-
-	if(target.status != LIGHT_OK)
-		if(CanUse(U))
-			if(!Use(U)) return
-			to_chat(U, "<span class='notice'>You replace the [target.fitting] with the [src].</span>")
-
-			if(target.status != LIGHT_EMPTY)
-
-				var/obj/item/weapon/light/L1 = new target.light_type(target.loc)
-				L1.status = target.status
-				L1.rigged = target.rigged
-				L1.brightness_range = target.brightness_range
-				L1.brightness_power = target.brightness_power
-				L1.brightness_color = target.brightness_color
-				L1.switchcount = target.switchcount
-				target.switchcount = 0
-				L1.update()
-
-				target.status = LIGHT_EMPTY
-				target.update()
-
-			var/obj/item/weapon/light/L2 = new target.light_type()
-
-			target.status = L2.status
-			target.switchcount = L2.switchcount
-			target.rigged = emagged
-			target.brightness_range = L2.brightness_range
-			target.brightness_power = L2.brightness_power
-			target.brightness_color = L2.brightness_color
-			target.on = target.has_power()
-			target.update()
-			del(L2)
-
-			if(target.on && target.rigged)
-				target.explode()
-			return
+	supply.remove_from_storage(best_light)
+
+	if(target.status != LIGHT_EMPTY)

```

Comment:
It would be much cleaner, but that's outside of the scope of this PR. The light fixture, as stupid as this is, does not actually contain a light, but rather just all the variables a light has.